pkg/hnsw/hnsw.go

```diff
@@ -207,30 +207,39 @@ func (h *Hnsw) InsertVector(q Point) error {
 		// add bidirectional connections from neighbors to q at layer c
 		for _, neighbor := range neighbors {
 			neighborPoint := h.points[neighbor.id]
-
 			distNeighToQ := EuclidDistance(*neighborPoint, q)
-
 			h.friends[neighbor.id].InsertFriendsAtLevel(level, qId, distNeighToQ)
 			h.friends[qId].InsertFriendsAtLevel(level, neighbor.id, distNeighToQ)
 		}
 
 		for _, neighbor := range neighbors {
 			neighborFriendsAtLevel, err := h.friends[neighbor.id].GetFriendsAtLevel(level)
-
 			if err != nil {
 				return fmt.Errorf("failed to find nearest neighbor to Q at level %v: %w", level, err)
 			}
 
-			maxNeighborsFriendsAtLevel := FromBaseQueue(neighborFriendsAtLevel, MaxComparator{})
+			maxNumberOfNeighbors := h.M
+			if level == 0 {
+				maxNumberOfNeighbors = h.Mmax0
+			}
+
+			eConnections := neighborFriendsAtLevel
+			if eConnections.Len() > maxNumberOfNeighbors {
+				var items []*Item
+
+				for !neighborFriendsAtLevel.IsEmpty() {
```

Comment:
This is likely going to be pretty slow. To speed it up, we probably want to use a min-max heap instead: https://en.wikipedia.org/wiki/Min-max_heap

This doesn't exist in the go std library but you can take inspiration from this library: https://github.com/esote/minmaxheap